src/gpio/gpio.c

```diff
@@ -61,68 +62,107 @@ mraa_gpio_init_internal(mraa_adv_func_t* func_table, int pin)
         status = dev->advance_func->gpio_init_pre(pin);
         if (status != MRAA_SUCCESS)
             goto init_internal_cleanup;
     }
-
     dev->value_fp = -1;
     dev->isr_value_fp = -1;
 #ifndef HAVE_PTHREAD_CANCEL
     dev->isr_control_pipe[0] = dev->isr_control_pipe[1] = -1;
 #endif
     dev->isr_thread_terminating = 0;
     dev->phy_pin = -1;
+    dev->owner = 1;
+    snprintf(dev->gpio_path, MAX_SIZE, "/gpio%d", dev->pin);
+    snprintf(bu, sizeof(bu), SYSFS_CLASS_GPIO "%s", dev->gpio_path);
 
-    // then check to make sure the pin is exported.
-    char directory[MAX_SIZE];
-    snprintf(directory, MAX_SIZE, SYSFS_CLASS_GPIO "/gpio%d/", dev->pin);
-    struct stat dir;
-    if (stat(directory, &dir) == 0 && S_ISDIR(dir.st_mode)) {
-        dev->owner = 0; // Not Owner
+    // check if the pin was already created, meaning another process may be the owner
+    if (stat(bu, &dir_stat) == 0 && S_ISDIR(dir_stat.st_mode)) {
+        dev->owner = 0;
     } else {
-        int export = open(SYSFS_CLASS_GPIO "/export", O_WRONLY);
-        if (export == -1) {
+        // old kernel gpio owner path names compatibility
+        for (i = 0; i < board->phy_pin_count; i++) {
+            if (board->pins[i].gpio.pinmap == pin) {
+                phy_pin = i;
+                break;
+            }
+        }
+
+        // gpio class creates paths all in lower case type
+        for (i = 0; i < MRAA_PIN_NAME_SIZE; i++)
+            pin_name[i] = tolower(board->pins[phy_pin].name[i]);
+
+        snprintf(dev->gpio_path, MAX_SIZE, "/gpio%d_%s", dev->pin, pin_name);
+        snprintf(bu, sizeof(bu), SYSFS_CLASS_GPIO "%s", dev->gpio_path);
+
+        // check if the pin was already created, meaning another process may be the owner
+        if (stat(bu, &dir_stat) == 0 && S_ISDIR(dir_stat.st_mode))
+            dev->owner = 0;
+    }
+
+    if (dev->owner == 1) {
+        export_fd = open(SYSFS_CLASS_GPIO "/export", O_WRONLY);
+        if (export_fd == -1) {
             syslog(LOG_ERR, "gpio%i: init: Failed to open 'export' for writing: %s", pin, strerror(errno));
             status = MRAA_ERROR_INVALID_RESOURCE;
             goto init_internal_cleanup;
         }
         length = snprintf(bu, sizeof(bu), "%d", dev->pin);
-        if (write(export, bu, length * sizeof(char)) == -1) {
+        if (write(export_fd, bu, length * sizeof(char)) == -1) {
             syslog(LOG_ERR, "gpio%i: init: Failed to write to 'export': %s", pin, strerror(errno));
-            close(export);
+            close(export_fd);
             status = MRAA_ERROR_INVALID_RESOURCE;
             goto init_internal_cleanup;
         }
-        dev->owner = 1;
-        close(export);
+
+        // check for an old or a new kernel version
+        snprintf(dev->gpio_path, MAX_SIZE, "/gpio%d_%s", dev->pin, pin_name);
+        snprintf(bu, sizeof(bu), SYSFS_CLASS_GPIO "%s", dev->gpio_path);
+        if (stat(bu, &dir_stat) == -1 || !S_ISDIR(dir_stat.st_mode))
```

Comment:
I think in mraa the recommended style is to enclose even one-line `if` to brackets